Add Dockerfile based on Ubuntu 20.04

gukandrew opened the pull request:
The official Dockerfile on alpine:3.11 doesn't work on Raspberypi OS (RPI4), this works fine

gukandrew committed:
Add Dockerfile based on Ubuntu 16.04

The official Dockerfile on alpine:3.11 doesn't work on Raspberypi OS (RPI4), this works fine

JabbaDesilijicTiure reviewed (changes requested):
The file server port is always random! 
The path should be ~/ and not **root**
**Tor never should be run as root!**

JabbaDesilijicTiure commented:
@imachug how you could approve this? :roll_eyes:

JabbaDesilijicTiure commented on Dockerfile.ubuntu:
```diff
@@ -0,0 +1,37 @@
+CMD (! ${ENABLE_TOR} || tor&) && python3 zeronet.py --ui_ip 0.0.0.0 --fileserver_port 26552
+
+#Expose ports
+EXPOSE 43110 26552
```
**You don't need to expose any port if you running it with Tor!** You could just allow in ufw:  `ufw allow in from yourclientip to serverip port 43110 proto tcp comment Zeronet`

gukandrew replied:
Why? I usually open ZN on host via 43110 port in browser :)

purplesyringa replied:
> You don't need to expose any port if you running it with Tor!

Correction: that is only true if you're running ZeroNet in Tor: Always mode. That might be a good option, but again: it's a separate feature request; the other dockerfiles do this too.

JabbaDesilijicTiure commented on Dockerfile.ubuntu:
```diff
@@ -0,0 +1,37 @@
+ pkg-config \
+ && python3 -m pip install --user --upgrade pip
+
+RUN pip3 install -r /root/requirements.txt \
```
```suggestion
RUN pip3 -m venv ~/zeroenv \
&& source ~/zeroenv/bin/activate \
&& pip3 install -r /root/requirements.txt \
```
Or something similar, need a virtual environment and as I said tor never should be run as root. You only need to modify the torrc file, create and than enable the tor service. 

At reboot your Tor will not work. Even possible to create a new user just for zeronet and so the path ~/ will be very useful. 

Neither Tor not ZeroNet should be run as root. 
Virtual environment

purplesyringa replied:
I'm not sure if it's a good idea -- ZeroNet is the only service running in this docker after all, so extra nesting is needless.

gukandrew replied:
I'm also not sure if this is really good idea, but I did that and it works.
So I can revert that and only run ZN and tor using non root user, or leave approach with virtuallenv.

JabbaDesilijicTiure commented:
@gukandrew You need to make significant changes. @imachug wrongfully approved your PR which is clearly dangerous.

gukandrew commented:
I'm not sure about that changes, as I took the [original Dockerfile](https://github.com/HelloZeroNet/ZeroNet/blob/py3/Dockerfile) as base. So that's why I have port and other things hardcoded as well as in original.
And you could use tor or not so you can set ENABLE_TOR env var.

Of course I can put your suggestions, but I think you can commit that directly (I have checked checkbox "Allow edits by maintainers").

gukandrew commented:
@JabbaDesilijicTiure @imachug  Btw why official docker image/dockerfile https://hub.docker.com/r/nofish/zeronet/dockerfile has same issues?

JabbaDesilijicTiure commented:
@gukandrew zeronet is not as good or secure as may you imagine. All suggestions made by the zeronet project like the dockerfile is outdated. @HelloZeroNet barely do anything than changing a rev number in Config.py. The only person who actually do something is @imachug. 

You should rewrite the entire dockerfile to be rootless: https://docs.docker.com/engine/security/rootless/ 
The docker file should only run a few commands as root like modifying torrc and enable it as a service. All other things including zeronet (should be venv) but at minimum should be on the path of a non-root user.

purplesyringa commented:
> The file server port is always random!

That's difficult to implement in Docker, and the other dockerfile in this repo has a hardcoded port as well. Making it random counts as a feature request, but you shouldn't ask the OP to do that.

purplesyringa commented:
> The only person who actually do something is @imachug.

Don't underestimate nofish who started this project and don't overestimate me who made a few contributions. We all burn out from time to time, it's okay.

JabbaDesilijicTiure commented:
@HelloZeroNet is a fascist living in EUs only fascist state in a full blown dictatorship. He cares nothing about ZeroNet, he maybe started the project but without contributors he would be nowhere and the only reason this project is not moving forward because he another authoritarian fascist managing the github repository. When people hear that ZeroNet is started by a Hungarian they immediately think about Orban Viktor and how the Hungarian police **beating and torturing refugees**. Tamas Kocsis or I should say @shortcutme is in fact proud that his country built in the south more than **520 km long electrified NATO razor fences**.

ZeroNet is not moving forward and never will move forward until Tamas is leading this project and publicizing Hungary with this software. The fact that Tamas was also subject of censorship by the TED in New York and he needed to rewrite his ZeroNet speach to talk about China, bad China who censors ZeroNet while he was totally silent about that in Hungary where he lives the **entire media printed and online is controlled by the full blown fascist government of Orban Viktor**. He did not even mentioned Hungary. He was just talking about communist China that how bad is it. 

@gukandrew @imachug

gukandrew commented:
@JabbaDesilijicTiure @imachug Wow I caught attention to more serious thing than I thought.
I'm not familiar with situation in Hungary, but if that is true it's terrible.
But in same time @HelloZeroNet is living in that country so maybe he is not talking about issues in Hungary because of concerns about close people life. IDK

JabbaDesilijicTiure commented:
@gukandrew yes it is all true. The main problem is he himself is censored not by the Hungarian but by the US government when he was giving a speech in New York about ZeroNet. All he did in the TED talk is talking nonsense things we already know, and a lot of shit about China without any mention of Hungary. 

**I do have nodes in China** and I can tell you @HelloZeroNet @imachug that **ZeroNet is not blocked in China**! Anyone who says the opposite is a fascist trash working either to the NATO or to some other International fascist organization. 

After the TED speech zeronet.io is blocked in China and this is only happened because the entire speech was about attacking China and glorifying the USA. Don't surprise when ZeroNet.io is banned there after such a speech. I bet that he just talked any shit you can about China without renting a few VPS in Beijing and testing ZeroNet himself. 

**Hypocrisy**

gukandrew commented:
I don't know a lot about China, but heard about Xinjiang re-education camps and The Great Firewall of China, that is also shows China not in good light (I do not justify the Hungary). Anyway the topic will not be fulfilled without mention that Russian Federation also is nazi country, that occupied Ukrainian Crimea and Eastern Donbas and asked support their actions for Hungary government at same time. So It would be hypocrisy to be silent about that from my side when topic goes about despotism. So what your thoughts about that fact?

@imachug @JabbaDesilijicTiure anyway I have fixed Dockerfile a bit, updated Ubuntu version and using virtualenv now.

JabbaDesilijicTiure commented:
Those camps in China are nothing more than NATO propaganda spread by the Americans here is a statement from United Nations officials who actually working for UN: https://thegrayzone.com/2018/08/23/un-did-not-report-china-internment-camps-uighur-muslims/

Russia? Without Russia we already dead body, we will be all either dead or live in brutal fascism. I see Russia as a savior, they did save Europe from the nazis one time and I strongly believe soon they actually go to war with NATO fascists in EU and save Europe again from fascists.

Crimea was legally annexed and I support that move. I don't like double standards, the Americans annexed entire countries in Afghanistan, Iraq, and they "protecting oil" now in Syria, Americans also "annexed" Guantanamo Bay which is a true concentration camp.  The real evil in this world is not China, Russia or Iran, the real evil and world's biggest terrorist which killing thousands of people with drones is the United States of America. I believe we are close to 3th world war and if that times come I volunteer to oust the Americans from Europe. 

About Hungary? Hungary is used by the Americans in the 90's as a HQ for troops on the way to Bosnia and Kosovo. Where the Americans actually involved in sex-trafficked, enslaved woman https://www.theguardian.com/world/2012/jan/15/bosnia-sex-trafficking-whistleblower and both in Bosnia and Kosovo the same companies involved in oil business which are today in Syria, Afghanistan and Iraq. 

Hungary joined NATO and immediately used by the Americans as a spot for future expansion. Orban Viktor is a result of the fascist NATO policies. The EU will never sanction  Orban or any fascists in the EU, they only sanction communists in Russia and Crimea. This whole thing with Navalny is also absolute lie. If any Russian official wanted to kill him he would be already dead. FSB would just walk up him and bum in the head. 

I think we just stop this conversation right here. I think @imachug actually lives in Russia. He could tell you what the CIA doing in Russia, Germany and elsewhere in Europe.

Regarding the dockerfile I open a review again because of the ports you expose unnecessary if Tor is used.

JabbaDesilijicTiure commented on Dockerfile.ubuntu:
```diff
@@ -0,0 +1,47 @@
+ENV ENABLE_TOR false
+
+#Set upstart command
+CMD (! ${ENABLE_TOR} || tor&) && . zeroenv/bin/activate && python3 zeronet.py --ui_ip 0.0.0.0 --fileserver_port 26552
```
```suggestion
CMD (! ${ENABLE_TOR} || tor&) && . zeroenv/bin/activate && python3 zeronet.py
```

./zeroenv/bin/activate maybe

gukandrew replied:
nope, dot is for "source"

gukandrew replied:
again, you couldn't proceed work with zeronet ui that run in container on your HOST in case when you haven't defined ports explicitly.

JabbaDesilijicTiure commented on Dockerfile.ubuntu:
```diff
@@ -0,0 +1,47 @@
+ && python3 -V \
+ && python3 -m pip list \
+ && tor --version \
+ && openssl version
```
. zeroenv/bin/activate \ 

you don't need this, I see it 3 different times in the dockerfile 
also python3 version and other versions don't need... Not important because you already in a virtual environment. you use pip3 and venv which all python3 higher than 3.4.

gukandrew replied:
Do you understand what is virtualenv and how dockerfile run action acts? Each run is like separate shell, you cannot activate that in first run and proceed calling binaries from activated env in second run. 
Why venv and not virtualenv? I'm actually not sure why do you force venv or virtualenv that adds new virtual layer in seperate docker container that only runs docker and tor?
Versions here is only for verbose info in logs.

JabbaDesilijicTiure commented on Dockerfile.ubuntu:
```diff
@@ -0,0 +1,47 @@
+FROM ubuntu:20.04
+
+#Timezone settings
+ENV TZ=Europe/Uzhgorod
```
Uzhgorod should be deleted. Add London instead. You don't need to expose where you live...

gukandrew replied:
For sure. It's not my town, but anyway.

JabbaDesilijicTiure commented on Dockerfile.ubuntu:
```diff
@@ -0,0 +1,47 @@
+CMD (! ${ENABLE_TOR} || tor&) && . zeroenv/bin/activate && python3 zeronet.py --ui_ip 0.0.0.0 --fileserver_port 26552
+
+#Expose ports
+EXPOSE 43110 26552
```
```suggestion

```

gukandrew replied:
I cannot do that, because after that I'll lose ability to open 127.0.0.1:43110 on my host.

gukandrew commented:
> Russia? Without Russia we already dead body, we will be all either dead or live in brutal fascism. I see Russia as a savior, they did save Europe from the nazis one time and I strongly believe soon they actually go to war with NATO fascists in EU and save Europe again from fascists.

Ha ha ha, they were nothing without Ukraine in WW2 as well as without USA's Lend-Lease.

> Crimea was legally annexed and I support that move. I don't like double standards, the Americans annexed entire countries in Afghanistan, Iraq, and they "protecting oil" now in Syria, Americans also "annexed" Guantanamo Bay which is a true concentration camp. The real evil in this world is not China, Russia or Iran, the real evil and world's biggest terrorist which killing thousands of people with drones is the United States of America. I believe we are close to 3th world war and if that times come I volunteer to oust the Americans from Europe.

Did you heard about Budapest Memorandum and the freezing of borders after World War II? They don't have any legal rights to occupy any territory of Ukraine. I hope the war between China and Russia will happen soon, and you'll feel the things that we in Ukraine feel now protecting our territory against russians nazis. I see real face of supporting terrorism from your side like Russian Federation actions related their neighbors. So after that I can't believe your justification of china camps and russian occupants.
And yes, I don't care about other countries after Russia invaded Ukraine why I should though?

We will be on opposite sides.

JabbaDesilijicTiure commented:
Do you have Telegram? Maybe we should continue this conversation there. 
Please fix the dockerfile. :smile:

gukandrew commented:
Nope, I really disappointed with your position. You're doing the same thing you claim @HelloZeroNet does. But in your case that clearly visible things, unlike situation with Tamas Kocsis.

JabbaDesilijicTiure commented:
Now it turns out that you are one of those Ukrainian neo-nazis financed by the U.S. State Department? :roll_eyes:

gukandrew commented:
lol, turn off russia24 on your tv. Any proofs? Except russian occupation of my country?

JabbaDesilijicTiure commented:
Your country run by a **clown** didn't you noticed? Neo-nazis waving EU flags and the whole government is working to get into the NATO. Russia should wipe out Ukraine from the map and if the Ukrainian state officials and NATO not step back you all will die by supersonic missiles. You speak Russian not German or French. and that invasion of Crimea is fake. Actually crimeans voted to be part of Russia. Fascist eurocentric russophobic maniacs should be gunned down and if this continues that way your country will be ground zeronet.

gukandrew commented:
> Your country run by a clown didn't you noticed?

Why you care who run our country? Fuck off.

> Neo-nazis waving EU flags and the whole government is working to get into the NATO

Why you care where other country goes? You have your China. So fuck off.

> Russia should wipe out Ukraine from the map

Fuck off, you're really terrorist.
And no, I speak Ukrainian. As well as another part of country. But why you care what language another country use?

> Actually crimeans voted to be part of Russia

It's fake claim, and fake referendum under the russian guns.

> Fascist eurocentric russophobic maniacs should be gunned down and if this continues that way your country will be ground zeronet.

It's not you who can decide that, and not another countries even. So fuck off again.

JabbaDesilijicTiure commented:
I reported you to the FSB, they know where you live you nazi pig.

gukandrew commented:
Fuck off again. And put FSB into your ass before that. NaziJabbaDesilijicTiure.

JabbaDesilijicTiure commented:
Long live Russia, China, Iran and all other countries standing up against fascist EU, NATO and American terrorists!

gukandrew commented:
I'm very scary. OMG.
@HelloZeroNet I cannot understand why this freak with nazi and terrorist views contribute this project.

JabbaDesilijicTiure commented:
You Ukrainian nazi pig! I hope Russian tanks smash your skull soon you shit.

gukandrew commented:
![image](https://user-images.githubusercontent.com/36536011/96336997-370d0c80-108c-11eb-9221-ea7049b4e57b.png)

purplesyringa commented:
@HelloZeroNet Please merge this PR and lock the discussion. And you guys please stop acting childish.

purplesyringa commented:
@gukandrew I am sorry for the insults. ZeroNet GitHub repository gets invaded by such people from time to time.

JabbaDesilijicTiure commented:
![nazi-ukraine](https://user-images.githubusercontent.com/71508279/96337031-50558f00-1073-11eb-8274-08ea74d723ce.jpeg)

Ukraine is not nazi at all. Sure, this is just Photoshop.

gukandrew commented:
@imachug thank you! I was almost going to close that PR because of that. That's very destructive behavior really will not provide good feedback for the project.

gukandrew commented:
@imachug sorry, one more thing. Question regarding virtualenv. Do we need that layer? As for me that would be overcomplicating, as we use our container only for ZN and could run as non root user even without virtualenv.

purplesyringa commented:
> As for me that would be overcomplicating, as we use our container only for ZN and could run as non root user even without virtualenv.

I think so too, perhaps we're not going to add more Python programs to the container.

gukandrew commented:
@imachug @HelloZeroNet dropped virtualenv. Now it should build faster.

gukandrew commented:
![Screenshot_2020-10-20 Add Dockerfile based on Ubuntu 20 04 by gukandrew · Pull Request #2652 · HelloZeroNet ZeroNet](https://user-images.githubusercontent.com/36536011/96570849-85f8b300-12d3-11eb-8218-e81f33b45466.jpg)

@JabbaDesilijicTiure you cannot erase your position as easy as you think.

scsmash3r commented:
@gukandrew Not regarding the topic and discussion about the actual subject, but I think that this Jabba person is just an ordinary troll, whose aim is to insult other people and to enjoy his self-affirmation (I was insulted by him too). Just ignore him and report the tech stuff.
